src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/WorkspaceProjectStateChangeDetector.cs

```diff
@@ -248,23 +249,26 @@ private void EnqueueUpdate(ProjectId projectId)
         {
             // A race is not possible here because we use the main thread to synchronize the updates
             // by capturing the sync context.
-            if (!_deferredUpdates.TryGetValue(projectId, out var update) || update.IsCompleted)
+            if (_deferredUpdates.TryGetValue(projectId, out var deferredUpdate) && !deferredUpdate.Task.IsCompleted && !deferredUpdate.Task.IsCanceled)
             {
-                _deferredUpdates[projectId] = UpdateAfterDelay(projectId);
+                deferredUpdate.Cts.Cancel();
             }
+
+            var cts = new CancellationTokenSource();
+            _deferredUpdates[projectId] = new UpdateItem(UpdateAfterDelay(projectId, cts), cts);
         }
 
-        private async Task UpdateAfterDelay(ProjectId projectId)
+        private async Task UpdateAfterDelay(ProjectId projectId, CancellationTokenSource cts)
         {
-            await Task.Delay(EnqueueDelay);
+            await Task.Delay(EnqueueDelay, cts.Token);
```

Comment:
If the token gets cancelled here it'll crash VS because nothing above it is catching it. You'll need to do:

```C#
try
{
    await Task.Delay(EnqueueDelay, cts.Token);
}
catch (TaskCanceledException)
{
    // Task cancelled, don't queue additional work.
    return;
}
```